File: app/src/main/java/com/example/aida64remote/data/Aida64SseClient.kt
package com.example.aida64remote.data

import com.example.aida64remote.model.ConnectionEvent
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.delay
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.isActive
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.util.concurrent.TimeUnit
import kotlin.math.min

class Aida64SseClient(
    private val client: OkHttpClient = defaultClient(),
) {
    fun connect(config: ConnectionConfig): Flow<ConnectionEvent> = flow {
        var attempt = 0
        var labels: Map<String, String> = emptyMap()

        while (currentCoroutineContext().isActive) {
            try {
                if (attempt == 0) {
                    emit(ConnectionEvent.Connecting)
                } else {
                    emit(ConnectionEvent.Reconnecting(attempt, null))
                }

                if (labels.isEmpty()) {
                    labels = fetchLabels(config)
                }

                val request = Request.Builder()
                    .url(config.sseUrl)
                    .header("Accept", "text/event-stream")
                    .header("Cache-Control", "no-cache")
                    .get()
                    .build()

                client.newCall(request).execute().use { response ->
                    if (!response.isSuccessful) {
                        throw IOException("HTTP ${response.code}")
                    }
                    val body = response.body ?: throw IOException("空响应")
                    emit(ConnectionEvent.Connected)
                    attempt = 0

                    BufferedReader(InputStreamReader(body.byteStream(), StandardCharsets.UTF_8)).use { reader ->
                        while (currentCoroutineContext().isActive) {
                            currentCoroutineContext().ensureActive()
                            val line = reader.readLine() ?: break
                            if (!line.startsWith("data:")) continue
                            val sensors = SensorParser.parse(line).map { item ->
                                item.copy(label = labels[item.id] ?: item.label)
                            }
                            if (sensors.isNotEmpty()) {
                                emit(ConnectionEvent.SensorsUpdated(sensors, labels))
                            }
                        }
                    }
                }

                throw IOException("SSE 连接已关闭")
            } catch (e: CancellationException) {
                throw e
            } catch (e: Exception) {
                attempt += 1
                val message = e.message ?: e.javaClass.simpleName
                emit(ConnectionEvent.Reconnecting(attempt, message))
                val delayMs = min(15_000L, (1L shl min(attempt - 1, 4)) * 1_000L)
                delay(delayMs)
            }
        }
    }.flowOn(Dispatchers.IO)

    private fun fetchLabels(config: ConnectionConfig): Map<String, String> {
        return try {
            val request = Request.Builder()
                .url(config.baseUrl + "/")
                .get()
                .build()
            client.newCall(request).execute().use { response ->
                if (!response.isSuccessful) return emptyMap()
                val html = response.body?.string().orEmpty()
                HtmlLabelParser.parseLabels(html)
            }
        } catch (_: Exception) {
            emptyMap()
        }
    }

    companion object {
        fun defaultClient(): OkHttpClient = OkHttpClient.Builder()
            .connectTimeout(10, TimeUnit.SECONDS)
            .readTimeout(0, TimeUnit.MILLISECONDS)
            .writeTimeout(10, TimeUnit.SECONDS)
            .retryOnConnectionFailure(true)
            .build()
    }
}

File: AIDA64Remote-iOS/AIDA64Remote/Data/Aida64SseClient.swift
import Foundation

actor Aida64SseClient {
    private let session: URLSession

    init(session: URLSession? = nil) {
        if let session {
            self.session = session
        } else {
            let config = URLSessionConfiguration.default
            config.timeoutIntervalForRequest = .infinity
            config.timeoutIntervalForResource = .infinity
            config.requestCachePolicy = .reloadIgnoringLocalCacheData
            config.waitsForConnectivity = false
            self.session = URLSession(configuration: config)
        }
    }

    func connect(config: ConnectionConfig) -> AsyncStream<ConnectionEvent> {
        AsyncStream { continuation in
            let task = Task {
                var attempt = 0
                var labels: [String: String] = [:]

                while !Task.isCancelled {
                    do {
                        if attempt == 0 {
                            continuation.yield(.connecting)
                        } else {
                            continuation.yield(.reconnecting(attempt: attempt, message: nil))
                        }

                        if labels.isEmpty {
                            labels = await fetchLabels(config: config)
                        }

                        var request = URLRequest(url: config.sseURL)
                        request.setValue("text/event-stream", forHTTPHeaderField: "Accept")
                        request.setValue("no-cache", forHTTPHeaderField: "Cache-Control")
                        request.timeoutInterval = .infinity

                        let (bytes, response) = try await session.bytes(for: request)
                        if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
                            throw URLError(.badServerResponse, userInfo: [
                                NSLocalizedDescriptionKey: "HTTP \(http.statusCode)",
                            ])
                        }

                        continuation.yield(.connected)
                        attempt = 0

                        for try await line in bytes.lines {
                            try Task.checkCancellation()
                            guard line.hasPrefix("data:") else { continue }
                            let sensors = SensorParser.parse(line).map { item in
                                var copy = item
                                copy.label = labels[item.id] ?? item.label
                                return copy
                            }
                            if !sensors.isEmpty {
                                continuation.yield(.sensorsUpdated(sensors: sensors, labels: labels))
                            }
                        }

                        throw URLError(.networkConnectionLost, userInfo: [
                            NSLocalizedDescriptionKey: "SSE 连接已关闭",
                        ])
                    } catch is CancellationError {
                        break
                    } catch {
                        attempt += 1
                        let message = error.localizedDescription
                        continuation.yield(.reconnecting(attempt: attempt, message: message))
                        let delayMs = Swift.min(15_000, (1 << Swift.min(attempt - 1, 4)) * 1_000)
                        try? await Task.sleep(nanoseconds: UInt64(delayMs) * 1_000_000)
                    }
                }

                continuation.finish()
            }

            continuation.onTermination = { _ in
                task.cancel()
            }
        }
    }

    private func fetchLabels(config: ConnectionConfig) async -> [String: String] {
        do {
            var request = URLRequest(url: config.baseURL)
            request.timeoutInterval = 10
            let (data, response) = try await session.data(for: request)
            if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
                return [:]
            }
            let html = String(data: data, encoding: .utf8) ?? ""
            return HtmlLabelParser.parseLabels(html)
        } catch {
            return [:]
        }
    }
}
